Upstream: No, upstream is dead unfortunately
From c72e431ba98990c34cce0b87231bd0092586955c Mon Sep 17 00:00:00 2001
From: Rasmus Thomsen <oss@cogitri.dev>
Date: Sat, 6 Jun 2020 12:37:12 +0200
Subject: [PATCH] Make /etc/localtime a symlink if it doesn't exist yet or
 already is a symlink

Previously we always tried to write the contents of the file in
/usr/share/zoneinfo to /etc/localtime, but that doesn't work if it
doesn't exist or if /etc/localtime is a symlink. As such, we should just
make it a symlink if it doesn't exist or already is one.
---
 src/timedated.c | 34 +++++++++++++++++++++++++++-------
 1 file changed, 27 insertions(+), 7 deletions(-)

diff --git a/src/timedated.c b/src/timedated.c
index 07c29da..a3402b1 100644
--- a/src/timedated.c
+++ b/src/timedated.c
@@ -139,13 +139,33 @@ set_timezone (const gchar *_timezone_name,
     localtime_filename = g_file_get_path (localtime_file);
     localtime2_filename = g_strdup_printf (DATADIR "/zoneinfo/%s", _timezone_name);
     localtime2_file = g_file_new_for_path (localtime2_filename);
-    if (!g_file_load_contents (localtime2_file, NULL, &filebuf, &length, NULL, error)) {
-        g_prefix_error (error, "Unable to read '%s':", localtime2_filename);
-        goto out;
-    }
-    if (!g_file_replace_contents (localtime_file, filebuf, length, NULL, FALSE, 0, NULL, NULL, error)) {
-        g_prefix_error (error, "Unable to write '%s':", localtime_filename);
-        goto out;
+
+    if (g_file_test(localtime_filename, G_FILE_TEST_IS_SYMLINK)) {
+        if (!g_file_delete (localtime_file, NULL, error)) {
+            g_prefix_error (error, "Unable to delete file to make new symlink %s:", localtime_filename);
+            goto out;
+        }
+        if (!g_file_make_symbolic_link (localtime_file, localtime2_filename, NULL, error)) {
+            g_error("Failed to make symlink %s", (*error)->message);
+            g_prefix_error (error, "Unable to create symlink %s -> %s:", localtime_filename, localtime2_filename);
+            goto out;
+        }
+    } else if (g_file_test(localtime_filename, G_FILE_TEST_IS_REGULAR)) {
+        if (!g_file_load_contents (localtime2_file, NULL, &filebuf, &length, NULL, error)) {
+            g_prefix_error (error, "Unable to read '%s':", localtime2_filename);
+            goto out;
+        }
+        if (!g_file_replace_contents (localtime_file, filebuf, length, NULL, FALSE, 0, NULL, NULL, error)) {
+            g_prefix_error (error, "Unable to write '%s':", localtime_filename);
+            goto out;
+        }
+    } else {
+        // File doesn't exist yet -> make a new symlink
+        if (!g_file_make_symbolic_link (localtime_file, localtime2_filename, NULL, error)) {
+            g_error("Failed to make symlink %s", (*error)->message);
+            g_prefix_error (error, "Unable to create symlink %s -> %s:", localtime_filename, localtime2_filename);
+            goto out;
+        }
     }
     ret = TRUE;
 
-- 
2.27.0

